### ISSUE TITLE
TypeError: cannot deepcopy this pattern object

### ISSUE THREAD

**Issue by leorich** [mesonbuild/meson#7971]
**Describe the bug**
With Meson 0.56 and the example below, I get a "TypeError: cannot deepcopy this pattern object" exception (the example in #7457 does *not* give an exception on this system with 0.56). If I rollback to 0.55 this example works as expected, but the example in #7457 throws the exception.

**To Reproduce**
meson.build

```
project('cmake_dep_regression','cpp')
cmake = import('cmake')

xtensor_dep = dependency('xtensor',
    version: ['>=0.21.4', '<1.0.0'],
    fallback: ['xtensor', 'xtensor_dep'],
    include_type: 'system',
    method : 'pkg-config')
exe1 = executable('main_xt', ['main_xt.cpp'], dependencies: [xtensor_dep])

# xerces_dep = cmake.subproject('xerces-c').dependency('xerces-c').as_system('system')
# exe1 = executable('main_cm', ['main_cm.cpp'], dependencies: [xerces_dep])
```

main_xt.cpp

``` cpp
#include <iostream>
#include <xtensor/xarray.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xview.hpp>

int main(void) {
  xt::xarray<double> arr1
      {{1.0, 2.0, 3.0},
       {2.0, 5.0, 7.0},
       {2.0, 5.0, 7.0}};

  std::cout << arr1 << std::endl;
  
  return 0;
}
```

xtensor.wrap

```
[wrap-file]
directory=xtensor-0.21.5

source_url=https://github.com/xtensor-stack/xtensor/archive/0.21.5.zip
source_filename=xtensor-0.21.5.zip
source_hash=f08aca41dc89128524dec04c82b590d9e0f1e923311e82bab9f024c99e7bb73a

patch_url=https://github.com/kyonifer/meson-wraps/raw/master/zips/xtensor-0.21.5-patches-1.zip
patch_filename=xtensor-0.21.5-patches-1.zip
patch_hash=d935d94a6b80d5466114cb29e49c2a358f42d0a1edceaa91f1ff76833b6c03e2
```

xtl.wrap

```
[wrap-file]
directory=xtl-0.6.12

source_url=https://github.com/xtensor-stack/xtl/archive/0.6.12.zip
source_filename=xtl-0.6.12.zip
source_hash=4ad3204312f72bce52951fb69acd1cd09e7f20f52c7e3b684140a4484acb5597

patch_url=https://github.com/kyonifer/meson-wraps/raw/master/zips/xtl-0.6.12-patches-2.zip
patch_filename=xtl-0.6.12-patches-2.zip
patch_hash=a65faee12aec22dd07f640863a0cae1b5310ea804e5d87c4a5ed73ee357ba6e1
```

nlohmann_json.wrap

```
[wrap-file]
directory = json-3.5.0

source_url = https://github.com/nlohmann/json/archive/v3.5.0.zip
source_filename = nlohmann_json-3.5.0.zip
source_hash = a65127f2797487cf07fdf1caacdff56ba3197ff02095cf0fde5bb35e91c86980

patch_url=https://github.com/kyonifer/meson-wraps/raw/master/zips/nlohmann_json-3.5.0-patches-1.zip
patch_filename=nlohmann_json-3.5.0-patches-1.zip
patch_hash=6585b48ef5cbf133d9c689c0b3748cc0ecb7324260ce87b96cf00d09bb0a1cdd
```

**Expected behavior**
successful build with both 0.55 and 0.56.

**system parameters**

Ubuntu 18.04.4 LTS
Python 3.6.9
meson 0.55.0 (works) meson 0.56.0 (fails)
ninja 1.8.2

**Comment by dcbaker** [comment]
Does this work without the `as_system` call?

**Comment by leorich** [comment]
I am not doing an `as_system` call that I know of - that commented out code in meson.build was from the other issue (in a separate test I uncommented it to check behavior of #7457 issue on my system).

There are some references to `as_system_method` in the traceback FWIW.

**Comment by dcbaker** [comment]
Oops, you're right, I didn't notice it was commented out

**Comment by dcbaker** [comment]
Could you post the full backtrace?

**Comment by leorich** [comment]
```
Using 'PKG_CONFIG_PATH' from environment with value: '/home/leo/repos/navigator_ws/devel/lib/pkgconfig:/opt/ros/melodic/lib/pkgconfig'
Using 'PKG_CONFIG_PATH' from environment with value: '/home/leo/repos/navigator_ws/devel/lib/pkgconfig:/opt/ros/melodic/lib/pkgconfig'
The Meson build system
Version: 0.56.0
Source dir: /home/leo/repos/test
Build dir: /home/leo/repos/test/build
Build type: native build
Project name: cmake_dep_regression
Project version: undefined
C++ compiler for the host machine: c++ (gcc 7.5.0 "c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C++ linker for the host machine: c++ ld.bfd 2.30
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: /usr/bin/pkg-config (0.29.1)
Run-time dependency xtensor found: NO (tried pkgconfig)
Looking for a fallback subproject for the dependency xtensor

|Executing subproject xtensor method meson 
|
|Project name: xtensor
|Project version: 0.21.5
|C++ compiler for the host machine: c++ (gcc 7.5.0 "c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
|C++ linker for the host machine: c++ ld.bfd 2.30
|Found CMake: /usr/bin/cmake (3.10.2)
|Using 'CMAKE_PREFIX_PATH' from environment with value: '/home/leo/repos/navigator_ws/devel:/opt/ros/melodic'
|Run-time dependency xtl found: NO (tried pkgconfig and cmake)
|Looking for a fallback subproject for the dependency xtl
|
||Executing subproject xtl method meson 
||
||Project name: xtl
||Project version: 0.6.12
||C++ compiler for the host machine: c++ (gcc 7.5.0 "c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
||C++ linker for the host machine: c++ ld.bfd 2.30
||Using 'CMAKE_PREFIX_PATH' from environment with value: '/home/leo/repos/navigator_ws/devel:/opt/ros/melodic'
||Run-time dependency nlohmann_json found: NO (tried pkgconfig and cmake)
||Looking for a fallback subproject for the dependency nlohmann_json
||
|||Executing subproject nlohmann_json method meson 
|||
|||Project name: nlohmann_json
|||Project version: 3.5.0
|||C++ compiler for the host machine: c++ (gcc 7.5.0 "c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
|||C++ linker for the host machine: c++ ld.bfd 2.30
|||Build targets in project: 0
|||Subproject nlohmann_json finished.
||
||Dependency nlohmann_json from subproject subprojects/json-3.5.0 found: YES 3.5.0
||Build targets in project: 0
||Subproject xtl finished.
|
|Dependency xtl from subproject subprojects/xtl-0.6.12 found: YES 0.6.12
|Dependency nlohmann_json found: YES 3.5.0 (cached)
|Build targets in project: 0
|Subproject xtensor finished.

Dependency xtensor from subproject subprojects/xtensor-0.21.5 found: YES 0.21.5
Traceback (most recent call last):
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/mesonmain.py", line 140, in run
    return options.run_func(options)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 253, in run
    app.generate()
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 161, in generate
    self._generate(env)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/msetup.py", line 200, in _generate
    intr.run()
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 4703, in run
    super().run()
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 519, in run
    self.evaluate_codeblock(self.ast, start=1)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 544, in evaluate_codeblock
    raise e
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 537, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 552, in evaluate_statement
    self.assignment(cur)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 1219, in assignment
    value = self.evaluate_statement(node.value)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 550, in evaluate_statement
    return self.function_call(cur)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 909, in function_call
    return func(node, func_args, kwargs)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 393, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 393, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 393, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  [Previous line repeated 3 more times]
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 209, in wrapped
    ret = f(*wrapped_args, **wrapped_kwargs)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 228, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 3734, in func_dependency
    d = d.as_system_method([wanted], {})
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 301, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreterbase.py", line 228, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/interpreter.py", line 516, in as_system_method
    new_dep = self.held_object.generate_system_dependency(new_is_system)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 235, in generate_system_dependency
    new_dep = copy.deepcopy(self)
  File "/usr/lib/python3.6/copy.py", line 161, in deepcopy
    y = copier(memo)
  File "/home/leo/.local/lib/python3.6/site-packages/mesonbuild/dependencies/base.py", line 261, in __deepcopy__
    setattr(result, k, copy.deepcopy(v, memo))
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 215, in _deepcopy_list
    append(deepcopy(a, memo))
  File "/usr/lib/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.6/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/usr/lib/python3.6/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/usr/lib/python3.6/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python3.6/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python3.6/copy.py", line 161, in deepcopy
    y = copier(memo)
TypeError: cannot deepcopy this pattern object
```

**Comment by drmoose** [comment]
It may be worth noting that this bug appears to be _specific_ to Python 3.6. Newer Pythons don't seem to reproduce it.

**Comment by borg323** [comment]
I have posted a minimal test case in https://github.com/mesonbuild/meson/issues/7920

**Comment by tristan957** [comment]
@eli-schwartz seems like this can be closed given the correction was merged in the linked ticket?

**Comment by eli-schwartz** [comment]
Yes, good catch. Fixed in #8216
